### PR TITLE
Restore precision setting for benchmark

### PR DESCRIPTION
**Pull request by ArtemySkrebkov-intel**
### Details:
 - It is a port of changes https://github.com/openvinotoolkit/openvino/pull/4167 without refactoring part not to touch other repos

### Tickets:
 - 51409

### CODE REVIEW

File: inference-engine/samples/benchmark_app/CMakeLists.txt [hunk 1]
```diff
@@ -9,4 +9,5 @@ ie_add_sample(NAME benchmark_app
               SOURCES ${SRC}
               HEADERS ${HDR}
               DEPENDENCIES format_reader
-              OPENCV_DEPENDENCIES imgcodecs)
+              OPENCV_DEPENDENCIES core)
```
Comment (ArtemySkrebkov-intel) [inline inference-engine/samples/benchmark_app/CMakeLists.txt:12]: Shouldn't be there.

File: inference-engine/samples/benchmark_app/inputs_filling.cpp [hunk 2]
```diff
@@ -145,7 +145,20 @@ void fillBlobBinary(Blob::Ptr& inputBlob,
 }
 
 template<typename T>
-void fillBlobRandom(Blob::Ptr& inputBlob) {
+using uniformDistribution =
```
Comment (ArtemySkrebkov-intel) [inline inference-engine/samples/benchmark_app/inputs_filling.cpp:148]: This file shouldn't be touched.

**Review by vinograd47 — APPROVED** [review]
Looks OK to me

File: inference-engine/samples/common/samples/args_helper.hpp [hunk 3]
```diff
@@ -85,3 +85,240 @@ void parseInputFilesArguments(std::vector<std::string> &files) {
         readInputFilesArguments(files, args.at(i));
     }
 }
+
+namespace {
+
+void splitStringList(const std::string& str, std::vector<std::string>& out, char delim) {
+    out.clear();
```
Comment (pelszkow) [inline inference-engine/samples/common/samples/args_helper.hpp:92]: Why this function doesn't return result? Is any reason why `out` is pass as parameter and clear it next line?

File: inference-engine/samples/common/samples/args_helper.hpp [hunk 4]
```diff
@@ -85,3 +85,240 @@ void parseInputFilesArguments(std::vector<std::string> &files) {
...
+
+    if (str.empty())
+        return;
+
+    std::istringstream istr(str);
+
+    std::string elem;
+    while (std::getline(istr, elem, delim)) {
```
Comment (pelszkow) [inline inference-engine/samples/common/samples/args_helper.hpp:100]: Consider split by `std::string::substr`.

File: inference-engine/samples/common/samples/args_helper.hpp [hunk 5]
```diff
@@ -85,3 +85,240 @@ void parseInputFilesArguments(std::vector<std::string> &files) {
...
+
+    std::map<std::string, std::string> parsedMap;
+    for (auto&& pair : pairs) {
+        std::vector<std::string> keyValue;
+        splitStringList(pair, keyValue, ':');
+        if (keyValue.size() != 2) {
+            throw std::invalid_argument("Invalid key/value pair " + pair + ". Expected <layer_name>:<value>");
+        }
+
+        parsedMap[keyValue[0]] = keyValue[1];
```
Comment (pelszkow) [inline inference-engine/samples/common/samples/args_helper.hpp:122]: Consider `std::map::insert`

File: inference-engine/samples/benchmark_app/benchmark_app.hpp [hunk 6]
```diff
@@ -237,4 +262,7 @@ static void showUsage() {
     std::cout << "    -load_config              " << load_config_message << std::endl;
 #endif
     std::cout << "    -qb                       " << gna_qb_message << std::endl;
+    std::cout << "    -ip                          <value>     "   << inputs_precision_message     << std::endl;
+    std::cout << "    -op                          <value>     "   << outputs_precision_message    << std::endl;
+    std::cout << "    -iop                        \"<value>\"    "   << iop_message                << std::endl;
```
Comment (pelszkow) [inline inference-engine/samples/benchmark_app/benchmark_app.hpp:267]: Do we want to flush each time?
```
Use of std::endl in place of '\n', encouraged by some sources, may significantly degrade output performance.
```
source :  https://en.cppreference.com/w/cpp/io/manip/endl

File: inference-engine/samples/benchmark_app/main.cpp [hunk 7]
```diff
@@ -67,6 +67,14 @@ bool ParseAndCheckCommandLine(int argc, char *argv[]) {
         throw std::logic_error("only " + std::string(detailedCntReport) + " report type is supported for MULTI device");
     }
 
+    bool isNetworkCompiled = fileExt(FLAGS_m) == "blob";
+    bool isPrecisionSet = !(FLAGS_ip.empty() && FLAGS_op.empty() && FLAGS_iop.empty());
+    if (isNetworkCompiled && isPrecisionSet) {
+        std::string err = std::string("Cannot set precision for a compiled network. ") +
+                          std::string("Please re-compile your network with required precision using compile_tool");
+
+        throw std::logic_error(err);
```
Comment (pelszkow) [inline inference-engine/samples/benchmark_app/main.cpp:76]: Why we need create string with message and not put directly literal to `logic_error` constructor?
```
logic_error( const char* what_arg );
```
source: https://en.cppreference.com/w/cpp/error/logic_error

**Comment by ArtemySkrebkov-intel** [comment]
Dear @pelszkow 
Thanks for your comments! This is a port of changes that it was already merged. Please see description.  I would like not to anything on top of that since it was already approved.

I suggest reviewing the following PR https://github.com/openvinotoolkit/openvino/pull/4318 instead where I port this changes into `master` branch and do a bit of refactoring for samples. We can address all of your comments in there. It would make much more sense.

**Review by ilyachur — APPROVED** [review]

**Pull request merged**